### ISSUE TITLE
How to create an RTCSessionDescription structure?

### ISSUE THREAD

**Issue by ineiti** [webrtc-rs/webrtc#155]
Hi there,

I'm trying to get `webrtc-rs` to interact with the browser-webrtc available in `web_sys.RtcPeerConnection`. I started with the example here:

https://github.com/webrtc-rs/examples/blob/5a0e2861c66a45fca93aadf9e70a5b045b26dc9e/examples/data-channels/data-channels.rs

So I'm creating an `APIBuilder::new()[...].new_peer_connection(config)` that should create an `offer`, which will be sent to the `web_sys.RtcPeerConnection` implementation.

The `web_sys` then sends back an `answer` that I would like to feed to the `peer_connection`, and here is my problem: `web_sys` only gives me a `String`, not an `RTCSessionDescription` as needed by the `peer_connection`. And the `RTCSessionDescription` has a private field, so I cannot do:

```
peer_connection.set_local_description(RTCSessionDescription{
  sdp_type: RTCSdpType::Answer,
  sdp: answer_string,
  parsed: None
}
```

I looked around at the methods of `peer_connection` but couldn't find an easy way to feed the answer-string to it.

**Comment by rainliu** [comment]
let desc_data = signal::decode(answer_string.as_str())?;
let answer = serde_json::from_str::<RTCSessionDescription>(&desc_data)?;

**Comment by ineiti** [comment]
OK, I'll try that. But to me it would be more logical to have something like:
```
impl RTCSessionDescription {
  fn offer(sdp: &str) -> Self {
    Self {
      sdp_type: RTCSdpType::Offer,
      sdp,
      parsed: None
    }
  }
}
```

And the same for `pranswer`, `answer`, ... So it would be clearer:
```
let answer = RTCSessionDescription::answer(answer_string);
```

Would that make sense?

**Comment by KallDrexx** [comment]
Adding my vote for this issue.  I had to add serde_json to my project just for this unfortunately, but it seems logical to me to be able to have a function convert a string SDP into a `RTCSessionDescription`.

**Comment by sax** [comment]
Same here. I'm interacting with webrtc.rs from Elixir, where I have the SDP as a string. It seems like the only mechanism I have to make a peer connection work is to generate JSON (which I have to guess the structure of) just so that it can be decoded, but that really doesn't make sense.

**Comment by rainliu** [comment]
I am not against adding new APIs to parse SDP string. 

PRs are welcome to address this issue.

**Comment by sax** [comment]
@ineiti I would be happy to issue a pull request, unless you already have code ready to go. I probably wouldn't be able to get to it until tomorrow.

**Comment by ineiti** [comment]
@sax - tomorrow sounds very fast. I'm currently counting weeks until I can get back to my side-project which uses webrtc :)

**Comment by sax** [comment]
Hahah. Well, maybe I was optimistic. I'll at least start it later today.

**Comment by sax** [comment]
@rainliu I was wondering if you might be able to point me in a direction. I'm writing the following test, and if fails because when I set the remote description on the answerer, it gives `Error: ErrSessionDescriptionMissingIceUfrag`. I don't see any mechanism to generate an offer that includes a ufrag or a password.

```rust
 #[tokio::test]
    async fn test_session_description_offer_answer() -> Result<()> {
        let mut m = MediaEngine::default();
        m.register_default_codecs()?;
        let api = APIBuilder::new().with_media_engine(m).build();

        let pc_offer = api.new_peer_connection(RTCConfiguration::default()).await?;
        let pc_answer = api.new_peer_connection(RTCConfiguration::default()).await?;

        let offer = pc_offer.create_offer(None).await?;
        let desc = RTCSessionDescription::offer(offer.sdp);

        pc_answer.set_remote_description(desc).await?;

        Ok(())
    }
```

Maybe I need to add something to the peer connection before generating the offer, like a data channel or something?

**Comment by sax** [comment]
Ok, I think I got it. Please take a look at the linked pull request. I'm happy to make changes if desired... I wasn't totally sure how to construct a realistic test for `pranswer` structs, so I basically just copied the `answer` test without the final assertion.

**Comment by jeremy-prater** [comment]
This might work? It seemed to work for me

```
let mut sdp = RTCSessionDescription::default();
sdp.sdp_type = RTCSdpType::Offer;
sdp.sdp = offer;
```

**Comment by k0nserv** [comment]
This was fixed in #167 so I'm closing this issue for now.
